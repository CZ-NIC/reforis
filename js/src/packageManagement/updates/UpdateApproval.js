/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";
import ReactTooltip from "react-tooltip";

import {
    useAPIPost,
    Button,
    Spinner,
    useAlert,
    API_STATE,
    Alert,
    ALERT_TYPES,
    toLocaleDateString,
    ForisURLs,
} from "foris";
import moment from "moment";
import API_URLs from "common/API";

UpdateApproval.propTypes = {
    update: PropTypes.object.isRequired,
    onSuccess: PropTypes.func.isRequired,
    delay: PropTypes.number,
};

export default function UpdateApproval({ update, onSuccess, delay }) {
    const [setAlert, dismissAlert] = useAlert();

    const [postState, post] = useAPIPost(API_URLs.approvals);
    // Execute callback when resolution is successful
    useEffect(() => {
        if (postState.state === API_STATE.ERROR) {
            setAlert(_("Cannot install updates."));
        } else if (postState.state === API_STATE.SUCCESS) {
            onSuccess();
            setAlert(
                _("Updates will be installed shortly."),
                ALERT_TYPES.SUCCESS
            );
        }
    }, [postState, onSuccess, setAlert]);

    function resolveUpdate(solution) {
        dismissAlert();
        post({ data: { hash: update.hash, solution } });
    }

    if (postState.state === API_STATE.SENDING) {
        return <Spinner className="text-center" />;
    }

    const packagesNumber =
        update.approvable &&
        babel.format(
            ngettext(
                "There is %d package to be updated.",
                "There are %d packages to be updated.",
                update.plan.length
            ),
            update.plan.length
        );
    const details = _(
        'See <a data-toggle="collapse" href="#plan-wrapper" role="button" aria-expanded="false" aria-controls="plan-wrapper">details</a>'
    );
    const delayedTime = moment(update.time).add(delay, "hour");
    return (
        <>
            {update.approvable ? (
                <>
                    {delay && (
                        <Alert type={ALERT_TYPES.SUCCESS}>
                            <span>
                                {_(
                                    `These updates are not going to be 
                                    automatically installed before 
                                    ${toLocaleDateString(delayedTime)}.`
                                )}
                            </span>
                            <i
                                className="fas fa-question-circle ml-1 help"
                                data-for="delay-tooltip"
                                data-tip={_(
                                    `If you don't want the updates to be installed at all, go to
                                    <a href="${ForisURLs.packageManagement.updateSettings}">Update Settings</a> 
                                    and choose the option <b>Update approval needed</b>.`
                                )}
                                data-event="click focus"
                                data-html
                            />
                            <ReactTooltip
                                effect="solid"
                                globalEventOff="click"
                                clickable
                                id="delay-tooltip"
                            />
                        </Alert>
                    )}
                    <h3>
                        {babel.format(
                            _("Approve Update From %s"),
                            toLocaleDateString(update.time)
                        )}
                    </h3>
                    <p
                        dangerouslySetInnerHTML={{
                            __html: `${packagesNumber} ${details}`,
                        }}
                    />
                    <div
                        className="collapse"
                        id="plan-wrapper"
                        data-testid="plan-wrapper"
                    >
                        <Plan plan={update.plan} />
                    </div>
                    <div className="text-right">
                        <Button
                            className="btn-primary"
                            onClick={() => resolveUpdate("grant")}
                            forisFormSize
                        >
                            {_("Install now")}
                        </Button>
                    </div>
                </>
            ) : (
                <p className="text-center text-muted">
                    {_("There are no updates awaiting your approval.")}
                </p>
            )}
        </>
    );
}

Plan.propTypes = {
    plan: PropTypes.arrayOf(PropTypes.object).isRequired,
};

function Plan({ plan }) {
    return (
        <div className="table-responsive">
            <table className="table table-hover">
                <thead className="thead-light">
                    <tr>
                        <th scope="col">{_("Action")}</th>
                        <th scope="col">{_("Name")}</th>
                        <th scope="col">{_("Current")}</th>
                        <th scope="col">{_("New")}</th>
                    </tr>
                </thead>
                <tbody>
                    {plan.map((item) => (
                        <PlanItem key={item.name} {...item} />
                    ))}
                </tbody>
            </table>
        </div>
    );
}

PlanItem.propTypes = {
    name: PropTypes.string.isRequired,
    op: PropTypes.string.isRequired,
    new_ver: PropTypes.string,
    cur_ver: PropTypes.string,
};

function PlanItem({ name, op, new_ver, cur_ver }) {
    return (
        <tr>
            <td>{op}</td>
            <td>{name}</td>
            <td>{cur_ver}</td>
            <td>{new_ver}</td>
        </tr>
    );
}
