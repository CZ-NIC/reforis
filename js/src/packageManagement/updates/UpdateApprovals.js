/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    useAPIGet, useAPIPost, Button, Spinner,
} from "foris";
import API_URLs from "common/API";
import toLocaleDateString from "utils/localeDate";

import isUpdateAvailable from "./utils";

export default function UpdateApprovals() {
    const [getState, get] = useAPIGet(API_URLs.approvals);
    const approval = getState.data;
    useEffect(() => {
        get();
    }, [get]);

    const [postState, post] = useAPIPost(API_URLs.approvals);
    useEffect(() => {
        get();
    }, [get, postState.data]);

    function postHandler(solution) {
        return (e) => {
            e.preventDefault();
            post({ hash: approval.hash, solution });
        };
    }

    if (getState.isLoading || postState.isSending) return <Spinner className="row justify-content-center" />;

    if (!isUpdateAvailable(approval)) {
        return <p>{_("There are no updates awaiting your approval.")}</p>;
    }

    const packagesNumber = babel.format(
        ngettext(
            "There is %d package to be updated.",
            "There are %d packages to be updated.",
            approval.plan.length,
        ),
        approval.plan.length,
    );
    const details = _(
        "See <a data-toggle=\"collapse\" href=\"#plan-wrapper\" role=\"button\" aria-expanded=\"false\" aria-controls=\"plan-wrapper\">details</a>",
    );

    const buttonMargin = { marginBottom: "1rem" };

    return (
        <>
            <h3>{babel.format(_("Approve update from %s"), toLocaleDateString(approval.time))}</h3>
            <p dangerouslySetInnerHTML={{ __html: `${packagesNumber} ${details}` }} />
            <div className="collapse" id="plan-wrapper" data-testid="plan-wrapper">
                <Plan plan={approval.plan} />
            </div>
            <Button
                style={buttonMargin}
                className="btn btn-warning offset-lg-1 col-lg-4 col-sm-12"
                onClick={postHandler("deny")}
            >
                {_("Deny")}
            </Button>
            <Button
                style={buttonMargin}
                className="btn btn-primary col-sm-12 col-lg-4 offset-lg-2 col-lg-3"
                onClick={postHandler("grant")}
            >
                {_("Install now")}
            </Button>
        </>
    );
}

Plan.propTypes = {
    plan: PropTypes.arrayOf(PropTypes.object).isRequired,
};

function Plan({ plan }) {
    return (
        <table className="table table-hover">
            <thead>
                <tr>
                    <th scope="col">{_("Action")}</th>
                    <th scope="col">{_("Name")}</th>
                    <th scope="col">{_("Current")}</th>
                    <th scope="col">{_("New")}</th>
                </tr>
            </thead>
            <tbody>
                {plan.map((item) => <PlanItem key={item.name} {...item} />)}
            </tbody>
        </table>
    );
}

PlanItem.propTypes = {
    name: PropTypes.string.isRequired,
    op: PropTypes.string.isRequired,
    new_ver: PropTypes.string,
    cur_ver: PropTypes.string,
};

function PlanItem({
    name, op, new_ver, cur_ver,
}) {
    return (
        <tr>
            <td>{op}</td>
            <td>{name}</td>
            <td>{cur_ver}</td>
            <td>{new_ver}</td>
        </tr>
    );
}
