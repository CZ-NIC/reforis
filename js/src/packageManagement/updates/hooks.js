/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import {
    API_STATE, useAlert, useAPIGet, useAPIPolling, useAPIPost,
} from "foris";
import { useCallback, useEffect, useState } from "react";

import API_URLs from "common/API";

const REFRESH_INTERVAL = 500; // milliseconds

/*
 * usePending is used to check/set if updater is running. It gets initial pending status from API.
 */
export function usePending() {
    const [pending, setPending] = useState(false);
    const [getUpdatesStatusResponse, getUpdatesStatus] = useAPIGet(API_URLs.updatesStatus);
    useEffect(() => {
        getUpdatesStatus();
    }, [getUpdatesStatus]);
    useEffect(() => {
        if (getUpdatesStatusResponse.state === API_STATE.SUCCESS) {
            setPending(getUpdatesStatusResponse.data.running);
        }
    }, [getUpdatesStatusResponse]);
    return [pending, setPending];
}

/*
 * useUpdateToApprove similar to useAPIGet(API_URLs.approvals). The getUpdateToApprove function
 * makes request only if displayApproval param is true. It makes initial request as well.
 */
export function useApprovals(displayApproval) {
    const [getApprovalsResponse, getApprovals] = useAPIGet(API_URLs.approvals);
    const updateToApprove = getApprovalsResponse.data;
    const getUpdateToApprove = useCallback(() => {
        // Request can be ignored if approval is hidden.
        if (displayApproval) {
            return getApprovals();
        }
    }, [displayApproval, getApprovals]);
    useEffect(() => {
        getUpdateToApprove();
    }, [getUpdateToApprove]);

    return [getApprovalsResponse, getApprovals, updateToApprove, getUpdateToApprove];
}

/*
 * usePendingPolling makes API polling until pending is false. It calls setPending to set result of
 * the polling.
 */
export function usePendingPolling(onSuccess, pending, setPending) {
    const [setAlert] = useAlert();

    const [checkStatusPollingResponse] = useAPIPolling(
        API_URLs.updatesStatus,
        REFRESH_INTERVAL,
        pending,
    );

    useEffect(() => {
        if (checkStatusPollingResponse.state === API_STATE.SUCCESS) {
            const isRunning = checkStatusPollingResponse.data.running;
            setPending(isRunning);
            if (!isRunning) {
                onSuccess();
            }
        } else if (checkStatusPollingResponse.state === API_STATE.ERROR) {
            setPending(false);
            setAlert(_("Cannot fetch updater status."));
        }
    },
    [checkStatusPollingResponse, onSuccess, setPending, setAlert]);
}

/*
 * useCheckUpdates is used to check for new updates. Sets pending as true when started.
 */
export function useCheckUpdates(setPending) {
    const [setAlert] = useAlert();

    const [runUpdatesResponse, runUpdates] = useAPIPost(API_URLs.runUpdates);
    // Trigger updater status checker after updater is enabled
    useEffect(() => {
        if (runUpdatesResponse.state === API_STATE.SUCCESS) {
            setPending(true);
        } else if (runUpdatesResponse.state === API_STATE.ERROR) {
            setPending(false);
            setAlert(runUpdatesResponse.data);
        }
    }, [runUpdatesResponse, setPending, setAlert]);

    return () => {
        setPending(true);
        runUpdates();
    };
}
