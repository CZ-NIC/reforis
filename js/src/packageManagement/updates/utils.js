export default function isUpdateAvailable(approval) {
    return (approval && approval.present && approval.status === "asked"
        && approval.update_automatically && approval.plan && approval.plan.length);
}
