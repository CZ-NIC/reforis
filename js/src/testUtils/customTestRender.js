/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
