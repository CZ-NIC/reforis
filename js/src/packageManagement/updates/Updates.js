import React from "react";

import UpdateApprovals from "./UpdateApprovals";

export default function Updates() {
    return (
        <>
            <h1>{_("Updates")}</h1>
            <UpdateApprovals />
        </>
    );
}
