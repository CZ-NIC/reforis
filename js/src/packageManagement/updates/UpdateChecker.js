/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { Button } from "foris";
import { useCheckUpdates, usePendingPolling } from "./hooks";

UpdateChecker.propTypes = {
    onSuccess: PropTypes.func.isRequired,
    pending: PropTypes.bool.isRequired,
    setPending: PropTypes.func.isRequired,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

export default function UpdateChecker({
    onSuccess, pending, setPending, children,
}) {
    usePendingPolling(onSuccess, pending, setPending);
    const refreshUpdates = useCheckUpdates(setPending);

    return (
        <Button
            className="btn-primary mb-3"
            forisFormSize
            onClick={refreshUpdates}
            disabled={pending}
        >
            {children}
        </Button>
    );
}
