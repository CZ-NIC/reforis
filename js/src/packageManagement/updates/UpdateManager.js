/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import {
    API_STATE, ErrorMessage, Spinner, withErrorMessage, withSpinnerOnSending,
} from "foris";
import UpdateChecker from "./UpdateChecker";
import UpdateApproval from "./UpdateApproval";
import { useApprovals, usePending } from "./hooks";

UpdateManager.propTypes = {
    displayChecker: PropTypes.bool.isRequired,
    checkerLabel: (props, propName) => {
        if (props.displayChecker === true && !props[propName]) {
            return new Error("checkerLabel is required if displayChecker is set to true");
        }
    },
    displayApproval: PropTypes.bool.isRequired,
    description: PropTypes.string,
};

function UpdateManager({
    displayChecker, checkerLabel, displayApproval, description,
}) {
    const [pending, setPending] = usePending();
    const [
        getApprovalsResponse,
        getApprovals,
        updateToApprove,
        getUpdateToApprove,
    ] = useApprovals(displayApproval);

    let approvalComponent;
    if (pending || getApprovalsResponse.state === API_STATE.SENDING) {
        approvalComponent = <Spinner className="text-center" />;
    } else if (getApprovalsResponse.state === API_STATE.ERROR) {
        approvalComponent = <ErrorMessage />;
    } else if (getApprovalsResponse.state === API_STATE.SUCCESS) {
        approvalComponent = (
            <UpdateApproval
                update={updateToApprove}
                onSuccess={getApprovals}
                className="mt-4"
            />
        );
    }

    return (
        <>
            {description}
            {displayChecker
            && (
                <UpdateChecker
                    onSuccess={getUpdateToApprove}
                    pending={pending}
                    setPending={setPending}
                >
                    {checkerLabel}
                </UpdateChecker>
            )}
            {approvalComponent}
        </>
    );
}

const UpdateManagerWithErrorAndSpinner = withSpinnerOnSending(withErrorMessage(UpdateManager));
export default UpdateManagerWithErrorAndSpinner;
