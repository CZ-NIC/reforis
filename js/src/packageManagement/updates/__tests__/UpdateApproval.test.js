/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, fireEvent, getByText, wait } from "foris/testUtils/customTestRender";
import mockAxios from 'jest-mock-axios';

import { mockJSONError } from "foris/testUtils/network";
import { mockSetAlert } from "foris/testUtils/alertContextMock";

import { exampleHash, exampleUpdate } from "./__fixtures__/updates";
import UpdateApproval from "../UpdateApproval";

describe("<UpdateApproval/>", () => {
    const onSuccess = jest.fn();

    function renderUpdateApproval(update=exampleUpdate) {
        const { container } = render(
            <UpdateApproval update={update} onSuccess={onSuccess} />
        );
        return container;
    }

    it("No updates awaiting", () => {
        const container = renderUpdateApproval({ ...exampleUpdate, approvable: false });
        expect(getByText(container, "There are no updates awaiting your approval.")).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    it("Updates awaiting - snapshot", () => {
        const container = renderUpdateApproval();
        expect(container).toMatchSnapshot();
    });

    it("Updates resolution - install now", async () => {
        const container = renderUpdateApproval();
        fireEvent.click(getByText(container, "Install now"));
        expect(mockAxios.post).toHaveBeenCalledWith("/api/approvals", {"hash": exampleHash, "solution": "grant"}, expect.anything());

        // Reload approvals when resolution is successful
        expect(onSuccess).not.toBeCalled();
        mockAxios.mockResponse({data: {}});
        await wait(() => expect(onSuccess).toBeCalled());
    });

    it("Updates resolution - ignore", () => {
        const container = renderUpdateApproval();
        fireEvent.click(getByText(container, "Ignore"));
        expect(mockAxios.post).toHaveBeenCalledWith("/api/approvals", {"hash": exampleHash, "solution": "deny"}, expect.anything());
    });

    it("Updates resolution - spinner", async () => {
        const container = renderUpdateApproval();
        fireEvent.click(getByText(container, "Install now"));
        expect(container).toMatchSnapshot();
    });

    it("Updates resolution - display error", async () => {
        const container = renderUpdateApproval();
        fireEvent.click(getByText(container, "Ignore"));
        mockJSONError();
        await wait(() => expect(mockSetAlert).toBeCalledWith("Cannot resolve update"));
    });
});
