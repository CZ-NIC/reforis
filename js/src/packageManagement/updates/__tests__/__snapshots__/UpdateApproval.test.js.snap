// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateApproval/> No updates awaiting 1`] = `
<div>
  <p
    class="text-center text-muted"
  >
    There are no updates awaiting your approval.
  </p>
</div>
`;

exports[`<UpdateApproval/> Updates awaiting - snapshot 1`] = `
<div>
  <div>
    <h3>
      Approve Update From %s
    </h3>
    <p>
      There is %d package to be updated. See 
      <a
        aria-controls="plan-wrapper"
        aria-expanded="false"
        data-toggle="collapse"
        href="#plan-wrapper"
        role="button"
      >
        details
      </a>
    </p>
    <div
      class="collapse"
      data-testid="plan-wrapper"
      id="plan-wrapper"
    >
      <table
        class="table table-hover"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Action
            </th>
            <th
              scope="col"
            >
              Name
            </th>
            <th
              scope="col"
            >
              Current
            </th>
            <th
              scope="col"
            >
              New
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              install
            </td>
            <td>
              PACKAGE
            </td>
            <td>
              1.0
            </td>
            <td>
              2.0
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <button
      class="btn btn-warning offset-lg-1 col-lg-4 col-sm-12"
      style="margin-bottom: 1rem;"
      type="button"
    >
       
       
      Ignore
    </button>
    <button
      class="btn btn-primary col-sm-12 col-lg-4 offset-lg-2 col-lg-3"
      style="margin-bottom: 1rem;"
      type="button"
    >
       
       
      Install now
    </button>
  </div>
</div>
`;

exports[`<UpdateApproval/> Updates resolution - spinner 1`] = `
<div>
  <div
    class="spinner-wrapper text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
